packages/openapi3/src/openapi.ts

```diff
@@ -424,20 +424,46 @@ function createOAPIEmitter(program: Program, options: OpenAPIEmitterOptions) {
     // Put them into currentEndpoint.responses
 
     for (const statusCode of statusCodes) {
-      if (currentEndpoint.responses[statusCode]) {
-        reportDiagnostic(program, {
-          code: "duplicate-response",
-          format: { statusCode },
-          target: responseModel,
-        });
-        continue;
-      }
-      const response: any = {
+      // the first model for this statusCode/content type pair carries the
+      // description for the endpoint. This could probably be improved.
+      const response = currentEndpoint.responses[statusCode] ?? {
         description: getResponseDescription(responseModel, statusCode),
       };
+
+      // check for duplicates
+      if (response.content) {
+        for (const contentType of contentTypes) {
+          if (response.content[contentType]) {
+            reportDiagnostic(program, {
+              code: "duplicate-response",
+              format: { statusCode, contentType },
+              target: responseModel,
+            });
+          }
+        }
+      }
+
       if (Object.keys(headers).length > 0) {
-        response.headers = headers;
+        response.headers ??= {};
+
+        // OpenAPI can't represent different headers per content type.
+        // So we merge headers here, and report any duplicates.
+        // It may be possible in principle to not error for identically declared
+        // yeaders.
+        for (const [key, value] of Object.entries(headers)) {
+          if (response.headers[key]) {
```

Comment:
Without ignoring headers that are the same how could one represent a common header or would it just not be possible yet.

Reply:
I don't think you could. But I didn't want to do much here in advance of the metadata proposal.

Reply:
(but I also felt that the story depending on declaring identical headers across all requests felt not so good, e.g. users might be surprised that headers from one content type affect headers for another content type).

Reply:
yeah that's true